uh50_api/response.py

```diff
@@ -0,0 +1,202 @@
+    "volume_previous_year_m3": {"regex": r"6.26\*01\((.*?)\*m3\)", "type": "float"},
+    "heat_previous_year_gj": {"regex": r"6.8\*01\((.*?)\*GJ\)", "type": "float"},
+    "error_number": {"regex": r"F\((.*?)\)", "type": "str"},
+    "device_number": {"regex": r"9.20\((.*?)\)", "type": "str"},
+    "measurement_period_minutes": {"regex": r"6.35\((.*?)\*m\)", "type": "int"},
+    "power_max_kw": {"regex": r"6.6\((.*?)\*kW\)", "type": "float"},
+    "power_max_previous_year_kw": {"regex": r"6.6\*01\((.*?)\*kW\)", "type": "float"},
+    "flowrate_max_m3ph": {"regex": r"6.33\((.*?)\*m3ph\)", "type": "float"},
+    "flowrate_max_previous_year_m3ph": {
+        "regex": r"6.33\*01\((.*?)\*m3ph\)",
+        "type": "float",
+    },
+    "flow_temperature_max_c": {"regex": r"9.4\((.*?)\*C", "type": "float"},
+    "return_temperature_max_c": {"regex": r"9.4\(.*?\*C&(.*?)\*C", "type": "float"},
+    "flow_temperature_max_previous_year_c": {
+        "regex": r"9.4\*01\((.*?)\*C",
+        "type": "float",
+    },
+    "return_temperature_max_previous_year_c": {
+        "regex": r"9.4\*01\(.*?\*C&(.*?)\*C",
+        "type": "float",
+    },
+    "operating_hours": {"regex": r"6.31\((.*?)\*h\)", "type": "int"},
+    "fault_hours": {"regex": r"6.32\((.*?)\*h\)", "type": "int"},
+    "fault_hours_previous_year": {"regex": r"6.32\*01\((.*?)\*h\)", "type": "int"},
+    "yearly_set_day": {"regex": r"6.36\((.*?)\)", "type": "str"},
+    "monthly_set_day": {"regex": r"6.36\*02\((.*?)\)", "type": "str"},
+    "meter_date_time": {
+        "regex": r"9.36\((.*?)\)",
+        "type": "datetime",
+        "format": "%Y-%m-%d&%H:%M:%S",
+    },
+    "measuring_range_m3ph": {"regex": r"9.24\((.*?)\*m3ph\)", "type": "float"},
+    "settings_and_firmware": {"regex": r"9.1\((.*?)\)", "type": "str"},
+    "flow_hours": {"regex": r"9.31\((.*?)\*h\)", "type": "int"},
+}
+
+
+@dataclass
+class HeatMeterResponse:
+    model: str
+    raw_response: str
+
+    # def __post_init__(self):
+    #     self._search_response(self.raw_response)
+
+    def _match(self, name):
+        str_match = re.search(
+            UH50_REGEX_CONFIG[name]["regex"], str(self.raw_response), re.M | re.I
+        )
+        if str_match:
+            try:
+                if UH50_REGEX_CONFIG[name]["type"] == "float":
+                    return float(str_match.group(1))
+                elif UH50_REGEX_CONFIG[name]["type"] == "int":
+                    return int(str_match.group(1))
+                elif UH50_REGEX_CONFIG[name]["type"] == "datetime":
+                    return datetime.datetime.strptime(
+                        str_match.group(1), UH50_REGEX_CONFIG[name]["format"]
+                    )
+                elif UH50_REGEX_CONFIG[name]["type"] == "str":
+                    return str_match.group(1)
+                else:
+                    return str_match.group(1)
+            except ValueError:
+                raise
+
+    @property

```

Comment:
Dit is inderdaad wat ik eerder suggereerde met die properties. Nu ik zie hoe je die class wilt gebruiken (het printen in main.py) zou ik dit toch nog weer anders opzetten. De dataclass geef je alle fields zoals ze moeten heten. Dat zorgt ervoor dat je in main.py met pprint in een keer een nette response kunt printen. Het uitpakken van die response (incl. de match methode en datastructuur) zou ik dan in een eigen class stoppen (HeatMeterResponseParser oid). Dan heb je het parsen van de response en het data object nog weer van elkaar gescheiden.

Reply:
`
    def parse_response(raw_response: str):
        heat_usage = _match("heat_usage_gj", raw_response)
        volume_usage = _match("volume_usage_m3", raw_response)
       `

Reply:
Zoiets bedoel ik dan.

Reply:
En nu ik mijn voorbeeldje nog eens bekijk, zit ook daar nog vrij veel duplicatie in. Het zou ook zo kunnen: 
![image](https://user-images.githubusercontent.com/3511418/170489744-1dcd4daa-d56f-407b-b725-b3904c42d7c0.png)

Reply:
Voordeel van deze opzet is dat wanneer er een nieuw veldje bij komt, je dat alleen maar in die config dictionary hoeft toe te voegen.

Reply:
which raises the question: waarom dan een dataclass. Dit is een afweging. Nadeel is dat je dus nog steeds op twee plekken die namen van variabelen hebt staan. Voordeel is dat je met typehints in je aanroepende code weet welke velden er zijn. Er zijn nog wel weer manieren om dat ook op te lossen hoor. Maar dat gaat dan misschien weer te ver voor nu.

Reply:
Zit ik hier zomaar mijn eigen review te reviewen.

Reply:
Gezien de hoeveelheid attributen is het wel lekker als je ze niet allemaal los hoeft te noemen (ook zonder pprint). Dan is een iterable oplossing wel lekker. En nu: TypeError: 'HeatMeterResponse' object is not iterable

Reply:
fields

Reply:
Ik ga dit inderdaad ombouwen. In de aanroepende kant heb ik het voor elkaar, nu in de service nog.

Reply:
Ik heb de eerste suggestie doorgevoerd. Het verder compact maken en ontdubbelen doe ik waarschijnlijk later nog. Even in stapjes die ik eerst goed zelf wil doorgronden.